File: src/docforge_sdk/docforge_sdk/models/transfers.py
# ====== Code Summary ======
# Response models for the collection transfer (export/import) endpoints, mirrored field-for-field
# from the DocForge backend router models (transfers/models.py): the 202 acceptance envelope and the
# full poll model a transfer's status is served from.

# ====== Standard Library Imports ======
from datetime import datetime
from typing import Any

# ====== Third-Party Library Imports ======
from pydantic import BaseModel, Field


class TransferAccepted(BaseModel):
    """
    The 202 envelope returned the instant a transfer is created and enqueued.

    Attributes:
        transfer_id (str): The tracking row's id; poll it for progress.
        kind (str): Transfer direction: "export" or "import".
        status (str): Lifecycle status at acceptance (pending).
    """

    transfer_id: str = Field(description="The tracking row's id; poll it for progress.")
    kind: str = Field(description="Transfer direction: 'export' or 'import'.")
    status: str = Field(description="Lifecycle status at acceptance (pending).")


class TransferStatus(BaseModel):
    """
    The full poll model of a collection transfer — its live status and (when done) its artifact.

    Attributes:
        transfer_id (str): The tracking row's id.
        kind (str): "export" or "import".
        status (str): pending / running / done / failed.
        progress (int): Coarse 0-100 percentage.
        stage (str | None): The current engine stage label, when running.
        counts (dict[str, Any] | None): Per-table counts snapshot ({"documents": n, "chunks": m, ...}).
        error (str | None): The failure message, verbatim, when failed.
        collection_id (str | None): Export → the source collection; import → the NEW collection once
            done (None while an import is still in flight).
        collection_name (str | None): The source (export) or produced (import) collection name.
        size_bytes (int | None): The produced export bundle's size (done export only).
        format_version (int | None): The bundle format version (done export only).
        dense_dim (int | None): The captured dense vector size (done export only).
        expires_at (datetime | None): When a produced export bundle may be garbage-collected.
        started_at (datetime | None): When the worker claimed the transfer.
        finished_at (datetime | None): When the transfer reached a terminal state.
        created_at (datetime): When the tracking row was created.
        updated_at (datetime): When the tracking row was last touched.
    """

    transfer_id: str = Field(description="The tracking row's id.")
    kind: str = Field(description="Transfer direction: 'export' or 'import'.")
    status: str = Field(description="Lifecycle status: pending / running / done / failed.")
    progress: int = Field(description="Coarse completion percentage (0-100).")
    stage: str | None = Field(default=None, description="Current engine stage label, when running.")
    counts: dict[str, Any] | None = Field(default=None, description="Per-table counts snapshot.")
    error: str | None = Field(default=None, description="Failure message (verbatim) when failed.")
    collection_id: str | None = Field(
        default=None, description="Export: source collection. Import: the new collection once done."
    )
    collection_name: str | None = Field(
        default=None, description="Source or produced collection name."
    )
    size_bytes: int | None = Field(
        default=None, description="Produced bundle size in bytes (done export)."
    )
    format_version: int | None = Field(
        default=None, description="Bundle format version (done export)."
    )
    dense_dim: int | None = Field(
        default=None, description="Captured dense vector size (done export)."
    )
    expires_at: datetime | None = Field(
        default=None, description="When a produced bundle may be GC'd."
    )
    started_at: datetime | None = Field(
        default=None, description="When the worker claimed the transfer."
    )
    finished_at: datetime | None = Field(
        default=None, description="When the transfer reached a terminal state."
    )
    created_at: datetime = Field(description="When the tracking row was created.")
    updated_at: datetime = Field(description="When the tracking row was last touched.")


__all__ = ["TransferAccepted", "TransferStatus"]

File: src/docforge/app/frontend/src/api/transfers.ts
// ====== Code Summary ======
// TypeScript mirror of the collection transfer (export/import) REST contract + its typed client.
// Shapes copied verbatim from the backend's Pydantic models
// (app/backend/routers/transfers/models.py) — nothing invented. The download helper mirrors
// api/blobs.ts: the route is authenticated, so a plain <a href> can't carry the Bearer header —
// the bytes are fetched behind auth and saved via an object URL instead.

import { apiFetch, apiFetchBlob } from "./http";

const COLLECTIONS_BASE = "/api/v1/collections";
const TRANSFERS_BASE = "/api/v1/transfers";

export type TransferKind = "export" | "import";
export type TransferStatusValue = "pending" | "running" | "done" | "failed" | string;

/** The 202 envelope returned the instant a transfer is created and enqueued. */
export interface TransferAccepted {
  transfer_id: string;
  kind: TransferKind;
  status: TransferStatusValue;
}

/** The full poll model of a collection transfer — its live status and (when done) its artifact. */
export interface TransferStatus {
  transfer_id: string;
  kind: TransferKind;
  status: TransferStatusValue;
  progress: number;
  stage: string | null;
  counts: Record<string, unknown> | null;
  error: string | null;
  /** Export: the source collection. Import: the new collection once done (null while in flight). */
  collection_id: string | null;
  collection_name: string | null;
  /** The produced bundle's size, in bytes — done export only. */
  size_bytes: number | null;
  format_version: number | null;
  dense_dim: number | null;
  /** When a produced bundle may be garbage-collected — done export only. */
  expires_at: string | null;
  started_at: string | null;
  finished_at: string | null;
  created_at: string;
  updated_at: string;
}

/** Open an asynchronous export of a whole collection into a portable `.dcexport` bundle. */
export function exportCollection(collectionId: string): Promise<TransferAccepted> {
  return apiFetch(`${COLLECTIONS_BASE}/${collectionId}/export`, { method: "POST" });
}

/** Import a `.dcexport` bundle as a brand-new collection (multipart upload). */
export function importCollection(file: File, targetName?: string): Promise<TransferAccepted> {
  const form = new FormData();
  form.append("file", file);
  if (targetName) form.append("target_name", targetName);
  return apiFetch(`${COLLECTIONS_BASE}/import`, { method: "POST", body: form });
}

/** Poll one transfer's live status — progress, stage, counts, error, and (when done) its artifact. */
export function getTransfer(transferId: string): Promise<TransferStatus> {
  return apiFetch(`${TRANSFERS_BASE}/${transferId}`);
}

/** Fetch a completed export bundle (authenticated) and trigger a browser download under `filename`. */
export async function downloadTransfer(transferId: string, filename: string): Promise<void> {
  const blob = await apiFetchBlob(`${TRANSFERS_BASE}/${transferId}/download`);
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  URL.revokeObjectURL(url);
}
